phylo/src/optimisers/branch_length_optimiser.rs

```diff
@@ -1,29 +1,113 @@
 }
 
-impl CostFunction for SingleBranchOptimiser<'_> {
+impl<'a, SubstModel: SubstitutionModel + Clone> CostFunction
+    for SingleBranchOptimiser<'a, SubstModel>
+where
+    SubstModel::ModelType: Clone,
+    SubstModel::Params: Clone,
+{
     type Param = f64;
     type Output = f64;
 
     fn cost(&self, value: &Self::Param) -> Result<Self::Output> {
-        let mut upd_tree = self.likelihood_cost.info.tree.clone();
-        upd_tree.set_branch_length(*self.branch_length, *value);
-        // self.likelihood_cost.info.tree = upd_tree;
-        Ok(-self.likelihood_cost.compute_log_likelihood(self.model).0)
+        let mut lc = PIPLikelihoodCost {
+            model: self.model,
+            info: self.phylo_info.clone(),
+        };
+        lc.info.tree.set_branch_length(self.branch, *value);
+        Ok(-lc.compute_log_likelihood().0)
     }
 
     fn parallelize(&self) -> bool {
         true
     }
 }
+
+pub struct BranchOptimiser<'a, SubstModel: SubstitutionModel> {
+    pub(crate) epsilon: f64,
+    pub(crate) model: &'a PIPModel<SubstModel>,
+    pub(crate) phylo_info: PhyloInfo,
+}
+
+impl<'a, SubstModel> BranchOptimiser<'a, SubstModel>
+where
+    SubstModel: SubstitutionModel + Clone,
+    SubstModel::ModelType: Clone,
+    SubstModel::Params: Clone,
+{
+    pub fn new(model: &'a PIPModel<SubstModel>, phylo_info: &PhyloInfo) -> Self {
+        BranchOptimiser {
+            epsilon: 1e-3,
+            model,
+            phylo_info: phylo_info.clone(),
+        }
+    }
+
+    pub fn optimise_parameters(&self) -> Result<(u32, Tree, f64, f64)> {
+        info!("Optimising branch lengths.");
+
+        let mut lc = PIPLikelihoodCost {
+            model: self.model,
+            info: self.phylo_info.clone(),
+        };
+
+        let init_logl = lc.compute_log_likelihood().0;
+        info!("Initial logl: {}.", init_logl);
+        let mut opt_logl = init_logl;
+        let mut prev_logl = f64::NEG_INFINITY;
+        let mut iters = 0;
+
+        let nodes: Vec<NodeIdx> = lc.info.tree.nodes.iter().map(|node| node.idx).collect();
+        while (prev_logl - opt_logl).abs() > self.epsilon {
+            debug!("Iteration: {}", iters);
+            prev_logl = opt_logl;
+            for branch in &nodes {
+                let (logl, length) = self.optimise_branch(&lc, branch)?;
+                opt_logl = logl;
+                lc.info.tree.set_branch_length(branch, length);
+                debug!(
+                    "Optimised branch length {} to value {} with logl {}",
+                    branch, length, opt_logl
+                );
+            }
+            iters += 1;
+        }
+        info!(
+            "Final logl: {}, achieved in {} iteration(s).",
+            opt_logl, iters
+        );
+        Ok((iters, lc.info.tree.clone(), init_logl, opt_logl))
+    }
+
+    fn optimise_branch(
+        &self,
+        lc: &PIPLikelihoodCost<SubstModel>,
+        branch: &NodeIdx,
+    ) -> Result<(f64, f64)> {
```

Comment:
ditto. probably a comment `// best_cost, best_param` would be good enough

Reply:
Will also be fixed in next PR, either with a comment or with something more meaningful.